Cython/Utility/ObjectHandling.c

```diff
@@ -528,10 +543,14 @@ static CYTHON_INLINE int __Pyx_SetItemInt_Fast(PyObject *o, Py_ssize_t i, PyObje
     if (is_list || PyList_CheckExact(o)) {
         Py_ssize_t n = (!wraparound) ? i : ((likely(i >= 0)) ? i : i + PyList_GET_SIZE(o));
         if ((!boundscheck) || likely(__Pyx_is_valid_index(n, PyList_GET_SIZE(o)))) {
+#if !CYTHON_AVOID_THREAD_UNSAFE_BORROWED_REFS
             PyObject* old = PyList_GET_ITEM(o, n);
-            Py_INCREF(v);
             PyList_SET_ITEM(o, n, v);
             Py_DECREF(old);
+#else
+            PyList_SetItem(o, n, v);
+#endif
+            Py_INCREF(v);
```

Comment:
Why did you move the incref to the end here? If `old == v`, I think we'd risk losing the reference.